src/lib/core/plugin.js

```diff
@@ -120,6 +121,11 @@ Plugin.prototype.registerContractsGeneration = function(cb) {
   this.addPluginType('contractGeneration');
 };
 
+Plugin.prototype.registerTestContractFactory = function(cb) {
+  this.testContractFactory = cb;
```

Comment:
So there's nothing to keep in mind and probably needs further discussion:

The current plugin API doesn't work very well for use cases like registering a contract factory for testing scenario. The reason for that is that there could be multiple plugins and use this API. Then the question is: which one is going to win? There can be only one factory function at a time.

The way it is done right now is that we always only take the last one that is registered (which could confuse people).

Another thing that wasn't clear to me, if this factory should affect both, `embark test` and `embark console`. 

I could imagine we probably don't want to blindly use the same factory for both scenarios, which is why in a follow up commit, there'd be a `registerConsoleContractFactory()` API. The same problems exist there as well though.

Over all this API addition feels like it needs more thought. Probably better done in an iteration where we rethink our entire Plugin API where needed.

Reply:
Yeah, the plugin API is really made so that the last one gets applied. It works well in the case of internal vs plugin, as the plugin will always win, which makes sense.

However, like you pointed, if there are multiple plugins that do the same thing, then it gets confusing.

We never had that issue before because the number of plugins is still small and no one tried to use two plugins that do the same thing at the same time (eg: two plugins to compile solidity files).

One thing to check is if the order in which we put the plugins in embark.json affects the order of application of the plugin, if so, it can be a workaround if we document it.